types/node-rules/index.d.ts

```diff
@@ -0,0 +1,58 @@
+declare class RuleEngine {
+  constructor();
+
+  /**
+   * Registers a rule with the rule engine.
+   * @param rule The rule to be registered.
+   */
+  register(rule: RuleEngine.Rule | RuleEngine.Rule[]): void;
+
+  /**
+   * Executes the rules on a given fact.
+   * @param fact The fact to be processed.
+   * @param callback The callback function to handle the result.
+   */
+  execute(fact: RuleEngine.Fact, callback: (data: RuleEngine.Fact & { result: any }) => void): void;
+}
+
+declare namespace RuleEngine {
+  interface Rule {
+    /**
+     * The condition to be evaluated for the rule.
+     * @param R The rule engine instance.
+     * @param fact The fact being evaluated.
+     */
+    condition: (R: FlowControlAPI, fact: Fact) => void;
+
+    /**
+     * The consequence to be executed if the condition is met.
+     * @param R The rule engine instance.
+     * @param fact The fact being evaluated.
+     */
+    consequence: (R: FlowControlAPI, fact: Fact) => void;
+
+    /**
+     * Optional parameter to specify the priority of the rule.
+     */
+    priority?: number;
```

Comment:
I highly suspect that this should all `| undefined` to support `exactOptionalPropertyTypes`.

See: [Playground Link](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2#code/PTAEHUEsBcAtQKYA8CGBjaB5ADtSB7AOxQBsAFAJ32wQugE8AVemgZ0WICMSEATAGlDMaAZTQVIuALAAoEKFgp2raBIy1QFAK492KKlsK9QAd0XRQ9fFtBoUhUKwQX7oaniKlZ87FRp16UGh8ADpZbzAAMXwKRFQAW2weQThIdkhCaFoAM3QEUF4ENBJ9BHY4FAs4Wny00BQI0F9qWgZTWEg0eDsHTnyifuygk3xHVQyAc1YALlAAcl59AGs5t1i5kkgJ2Gg5xpjQGEdYaxJjQnwLPsOHardOACsi6DCZWQysily0fIBVJwoABEELkdNB2ABvWSgUDyRiwfIoThOQg-NxDFCgABupC0+QoCF8ZQQmXYc1Y9BUCHiexkMLQ+BIMXh1IQmCxtAkhQA-LMAESLChLPmgAA+oD5m220D5AG5ZABfcJyMBQODWKoddjZEoTDhIngCIIIgn1U1wAn5HEkPHsMydeBWGw9RpOCwAAwZTIoLPibI5FC5CHdQVGfX5guFgklWx2IvsxndhkK2QyfHdr0aImceEIeru1rxodASaMILTvBDJkgJBI9Vr+BMoHi+BUoAAUigcWIJLhNIY8H7Gl0ikttQc7sg0llUflgqBcjXBPauod2CDss9IBySIFciQKa95AAJRsIAMpE0Icn1BdaABe97TrHKo1S7D6iixBAoISECIKctCBgAhCEabIDk9RlmQRP12U5SBChDAB9Xh8DKQg5mgZCbmKLRChLZMgPTTMVX-fIdRQPV3WQdAsFwUDSEoFoAmEMoQ3iFAljKIQWAQHtJBcNA0C0ChKgQXdGk46BVzuFMMhAogmiob9CmMJR6gcdxGLrZp-AYUjCmKUo71RDwHAmZx-lobNoFzKYAAoAEpZmsoEQRQMFWHlN4ZAZQg2zdez2AAXlASzoDc2zgucnygsmVgQi9GDWXgwNEPyMKBWWOVZHivNEuSn1YP9BCCKyqU4x8xoACEtAsC5oGmPKcwSpLoOK1KAyDUAwqI1NCD4aqyMoH9izuSjqNojAcHMpi-FaJg+NYd1LzqEwNMappW1YSBuAQEIgA)